captum/optim/_utils/reducer.py

```diff
@@ -0,0 +1,112 @@
+from typing import Any, Callable, List, Union
+
+import numpy as np
+import torch.nn.functional as F
+
+try:
+    import sklearn.decomposition
+    from sklearn.base import BaseEstimator
+except (ImportError, AssertionError):
+    print(
+        "The sklearn library is required to use Captum's ChannelReducer"
+        + " unless you supply your own reduction algorithm."
+    )
+import torch
+
+
+class ChannelReducer:
+    """
+    Dimensionality reduction for the channel dimension of an input.
+
+    Olah, et al., "The Building Blocks of Interpretability", Distill, 2018.
+    See: https://distill.pub/2018/building-blocks/
+    """
+
+    def __init__(
+        self, n_components: int = 3, reduction_alg: Any = "NMF", **kwargs
+    ) -> None:
+        if isinstance(reduction_alg, str):
+            reduction_alg = self._get_reduction_algo_instance(reduction_alg)
+            if reduction_alg is None:
+                raise ValueError(
+                    "Unknown sklearn dimensionality reduction method '%s'."
+                    % reduction_alg
+                )
+
+        self.n_components = n_components
+        self._reducer = reduction_alg(n_components=n_components, **kwargs)
+
+    def _get_reduction_algo_instance(self, name: str) -> Union[None, Callable]:
+        if hasattr(sklearn.decomposition, name):
+            obj = sklearn.decomposition.__getattribute__(name)
+            if issubclass(obj, BaseEstimator):
+                return obj
+        return None
+
+    @classmethod
+    def _apply_flat(cls, func: Callable, x: torch.Tensor) -> torch.Tensor:
+        orig_shape = x.shape
+        return func(x.reshape([-1, x.shape[-1]])).reshape(list(orig_shape[:-1]) + [-1])
+
+    def fit_transform(
+        self, x: torch.Tensor, swap_2nd_and_last_dims: bool = True
+    ) -> torch.Tensor:
+        """
+        Perform dimensionality reduction on an input tensor or NumPy array.

```

Comment:
nit: we might need to update the description about numpy array

Reply:
I've removed the part about NumPy arrays.